### PR TITLE
Defer CI reruns until workflows complete

### PR DESCRIPTION
**Pull request by subsetpark**
## Summary

- poll a GitHub Actions workflow run until it completes before requesting a failed-job rerun
- deduplicate failed checks so each workflow run receives only one rerun request
- treat an already-running rerun won by another actor as success
- add pure property coverage for status decoding, workflow URL parsing, and deduplication

## Testing

- `dune build`
- `dune runtest`
- `dune fmt`

---
<a href="https://app.blacksmith.sh/flowglad/codesmith/onton/pr/419"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1789752360&installation_model_id=19519&pr_number=419&repository=flowglad%2Fonton&return_to=https%3A%2F%2Fgithub.com%2Fflowglad%2Fonton%2Fpull%2F419&signature=41ded9ed0251080c6b386b8356959610103b1934625a51eab9fe4079d995cbb6"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved rerunning of failed CI jobs by waiting for workflow runs to finish before submitting rerun requests.
  - Prevented duplicate rerun requests for checks tied to the same workflow run.
  - Treats already-running workflows as successful rerun requests while preserving other error reporting.
  - Improved handling of incomplete, malformed, or unrecognized workflow responses without unexpected failures.

- **Tests**
  - Added coverage for workflow status handling, URL parsing, deduplication, and edge cases.

### CODE REVIEW

File: lib/github.ml [hunk 1]
```diff
@@ -1331,12 +1318,48 @@ let rerun_failed_jobs_for_check ~net ~clock ?timeout t
...
+        | Error _ as error -> error
+        | Ok body -> (
+            match Ci_rerun_decision.workflow_status_of_response body with
+            | Completed -> request_rerun ()
+            | Pending ->
+                Eio.Time.sleep clock 10.0;
+                wait_until_complete ()
+            | Malformed ->
+                Error
+                  (Json_parse_error
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/github.ml:1345]: Fixed. The workflow-status polling phase is now wrapped in an overall `Eio.Time.with_timeout` using the supplied timeout (or the existing default), so the deadline covers every GET and the 10-second sleeps. Exhausting that budget returns the existing `Github.Timeout` error; the rerun POST remains a separately bounded request.

File: lib_core/ci_rerun_decision.ml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  | Some pos ->
+      let start = pos + String.length marker in
+      let rest = String.drop_prefix url start in
+      let len =
+        match String.findi rest ~f:(fun _ ch -> not (Char.is_digit ch)) with
+        | None -> String.length rest
+        | Some (idx, _) -> idx
+      in
+      let id = String.prefix rest len in
+      if String.is_empty id then None else Int.of_string_opt id
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib_core/ci_rerun_decision.ml]: Fixed. `workflow_run_id_from_url` now accepts the digit segment only when it ends at `/`, `?`, `#`, or end-of-string. Invalid suffixes such as `42x` return `None`, with boundary coverage for all valid terminators and the invalid suffix.

File: lib/github.ml [hunk 3]
```diff
@@ -1331,12 +1318,48 @@ let rerun_failed_jobs_for_check ~net ~clock ?timeout t
...
-        ~f:(fun _ -> ())
+      let rec wait_until_complete () =
+        match request ~net ~clock ?timeout t ~meth:`GET ~path:run_path () with
+        | Error _ as error -> error
+        | Ok body -> (
+            match Ci_rerun_decision.workflow_status_of_response body with
+            | Completed -> request_rerun ()
+            | Pending ->
+                Eio.Time.sleep clock 10.0;
+                wait_until_complete ()
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/github.ml:1342]: Fixed. `wait_until_complete` now has one aggregate timeout across all status requests and sleeps, rather than restarting the budget for each GET. A still-pending workflow at the deadline returns the existing timeout error and releases the caller's session slot.

File: lib/github.ml [hunk 4]
```diff
@@ -1331,12 +1318,48 @@ let rerun_failed_jobs_for_check ~net ~clock ?timeout t
...
+            | Malformed ->
+                Error
+                  (Json_parse_error
+                     (Printf.sprintf
+                        "workflow run %d response has no valid status" run_id)))
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/github.ml:1347]: Fixed. The workflow-status constructors are now qualified with `Ci_rerun_decision` in both `lib/github.ml` and the property tests. This also allowed removal of the warning-40/42 suppression; the target compiles under the normal fatal-warning policy.

File: test/dune [hunk 5]
```diff
@@ -58,6 +58,13 @@
  (modules test_github_check_details)
  (libraries onton onton_core base yojson))
 
+(test
+ (name test_ci_rerun_decision_properties)
+ (modules test_ci_rerun_decision_properties)
+ (libraries onton_core qcheck-core qcheck-core.runner base)
+ (ocamlopt_flags (:standard -w -40-42))
+ (ocamlc_flags (:standard -w -40-42)))
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/dune]: Fixed. Removed the target-specific `ocamlopt_flags` and `ocamlc_flags` that suppressed warnings 40 and 42. The remaining constructor and record-field references are explicitly qualified, and `dune build` passes with the standard fatal-warning policy.

File: test/test_ci_rerun_decision_properties.ml [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        true
+      with _ -> false)
+
+let prop_dedupe_total =
+  QCheck2.Test.make ~name:"workflow check dedupe is total" ~count:1_000
+    gen_checks (fun checks ->
+      try
+        let _ = Ci_rerun_decision.unique_workflow_checks checks in
+        true
+      with _ -> false)
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_ci_rerun_decision_properties.ml]: Fixed. The three totality properties now call the pure decision functions directly and return `true`, without broad exception handlers, so any unexpected exception fails the test visibly.

File: test/test_ci_rerun_decision_properties.ml [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+      Option.equal Int.equal
+        (Ci_rerun_decision.workflow_run_id_from_url
+           "https://github.com/o/r/actions/runs/42/job/7")
+        (Some 42)
+      && Option.is_none
+           (Ci_rerun_decision.workflow_run_id_from_url
+              "https://github.com/o/r/actions/runs/")
+      && Option.is_none
+           (Ci_rerun_decision.workflow_run_id_from_url
+              "https://github.com/o/r/actions/runs/not-a-number/job/7"))
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_ci_rerun_decision_properties.ml]: Fixed. Added a generated interleaving property that mixes repeated workflow IDs with unidentified checks and compares the result against first-occurrence retention in original order. The URL boundary property now also verifies that an overflowing run ID returns `None` (along with invalid-suffix and valid-terminator cases).

File: lib/github.ml [hunk 8]
```diff
@@ -1331,12 +1318,48 @@ let rerun_failed_jobs_for_check ~net ~clock ?timeout t
...
-      Result.map (request ~net ~clock ?timeout t ~meth:`POST ~path ())
-        ~f:(fun _ -> ())
+      let rec wait_until_complete () =
+        match request ~net ~clock ?timeout t ~meth:`GET ~path:run_path () with
+        | Error _ as error -> error
+        | Ok body -> (
+            match Ci_rerun_decision.workflow_status_of_response body with
+            | Completed -> request_rerun ()
+            | Pending ->
+                Eio.Time.sleep clock 10.0;
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/github.ml:1341]: Addressed in later commit `7082757`. The polling phase now uses one overall `Eio.Time.with_timeout` budget across every status GET and 10-second sleep, returning the existing `Github.Timeout` error if the workflow remains non-terminal. The rerun POST remains separately bounded.

File: test/test_ci_rerun_decision_properties.ml [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        (count_unidentified checks)
+        (count_unidentified (Ci_rerun_decision.unique_workflow_checks checks)))
+
+let prop_status_boundaries =
+  QCheck2.Test.make ~name:"only a completed workflow status permits a rerun"
+    ~count:1 QCheck2.Gen.unit (fun () ->
+      Ci_rerun_decision.equal_workflow_status
+        (Ci_rerun_decision.workflow_status_of_response
+           {|{"status":"completed"}|})
+        Completed
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_ci_rerun_decision_properties.ml]: Addressed in later commit `7082757`. The test now uses `Ci_rerun_decision.Completed`, `Ci_rerun_decision.Pending`, and `Ci_rerun_decision.Malformed`; the warning-40/42 suppression was also removed, and the target passes under the normal fatal-warning policy.

File: lib_core/ci_rerun_decision.ml [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  | None -> None
+  | Some pos ->
+      let start = pos + String.length marker in
+      let rest = String.drop_prefix url start in
+      let len =
+        match String.findi rest ~f:(fun _ ch -> not (Char.is_digit ch)) with
+        | None -> String.length rest
+        | Some (idx, _) -> idx
+      in
+      let id = String.prefix rest len in
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib_core/ci_rerun_decision.ml]: Addressed in later commit `7082757`. The parser now accepts a numeric run ID only when the digit segment ends at `/`, `?`, `#`, or end-of-string, so digit-prefixed invalid segments such as `42x` return `None`.

File: test/test_ci_rerun_decision_properties.ml [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+              {|{"status":"in_progress"}|})
+           Pending
+      && Ci_rerun_decision.equal_workflow_status
+           (Ci_rerun_decision.workflow_status_of_response {|{"status":null}|})
+           Malformed
+      && Ci_rerun_decision.equal_workflow_status
+           (Ci_rerun_decision.workflow_status_of_response "not-json")
+           Malformed)
+
+let prop_url_boundaries =
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_ci_rerun_decision_properties.ml:151]: Addressed in later commit `7082757`. `prop_url_boundaries` now includes a 128-digit run ID and verifies that integer overflow returns `None`.

File: test/test_ci_rerun_decision_properties.ml [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+let prop_identifiable_runs_unique =
+  QCheck2.Test.make
+    ~name:"at most one rerun request is planned for each workflow run"
+    ~count:1_000 gen_checks (fun checks ->
+      let ids =
+        Ci_rerun_decision.unique_workflow_checks checks
+        |> List.filter_map ~f:run_id
+      in
+      let unique_ids = Set.of_list (module Int) ids in
+      List.length ids = Set.length unique_ids)
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_ci_rerun_decision_properties.ml:120]: Addressed in later commit `7082757`. The generated `prop_dedupe_interleavings` property mixes repeated workflow IDs with unidentified checks and verifies exact first-occurrence retention and original input order.

File: lib/github.ml [hunk 13]
```diff
@@ -1330,13 +1317,59 @@ let rerun_failed_jobs_for_check ~net ~clock ?timeout t
...
+      let run_path =
+        Printf.sprintf "/repos/%s/%s/actions/runs/%d" t.owner t.repo run_id
+      in
+      let rerun_path =
         Printf.sprintf "/repos/%s/%s/actions/runs/%d/rerun-failed-jobs" t.owner
           t.repo run_id
       in
-      Result.map (request ~net ~clock ?timeout t ~meth:`POST ~path ())
-        ~f:(fun _ -> ())
+      let polling_timeout = Option.value timeout ~default:default_timeout in
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/github.ml]: Fixed. The aggregate workflow-completion deadline is now a separate 30-minute budget instead of inheriting the 30-second per-request timeout. Individual GitHub API calls still fail quickly after 30 seconds, while normally queued or multi-minute workflows can reach a terminal state and trigger the rerun request; the overall poll remains bounded so it cannot retain the session slot forever. `dune build` and the full `dune runtest` suite pass.
